### ISSUE TITLE
Fixing the JS/CSS situation

### ISSUE THREAD

**Issue by abbysmal** [Engil/Canopy#17]
At the moment, the CSS/JS situation is kind of bad.
We got a few CSS files and JS files we need to retrieve via Bower (which isn't good), and load them somehow in the templates.
@wiredsister did already a great job (thank you!) at cleaning the Less files and making them easier to work on, but here is a list of the thing missing I think I'll be working on next, following @wiredsister suggestions:
- [x] Use browserify to bundle each JS libraries into one: simpler loading, simpler embedding is needed
- [x] Try to get rid of style.sh and use config.ml to handle that: on the #mirage IRC, drup suggested to write a custom device to allow running arbitrary shell commands within config.ml. I can't find any example of such a thing and I'm not sure this is actually better than the current situation
- [x] Get rid of bower: Remove it and use npm and browserify to handle JS libraries, halfway done somewhere in one of my local branches
- [ ] Allow simple template customization through Less files: by using some variable, allowing users to tweak a little bit of their canopy instance look: nothing fancy, but things like colors would a least be nice. That's what I was targeting at first when writing it, but I have a few ideas on how to do it so I'll be working on it one day.
- [x] Embed JS and CSS file into the repository (?): I'm not sure on how to this, but the main rationale being it would be avoiding the npm and less dependencies if the user don't want to tweak the interface.

I think I covered most of the things related to CSS/JS, if you have suggestions on how to properly do most of those things, those are welcome. :)
(also ping @hannesm, @dinosaure)

**Comment by hannesm** [comment]
I'm in favour of the last item: include the CSS and JS in the repository (but retain the possibility to manually fetch and install those things via npm), or as an external resource (tarball somewhere).  Thus to install Canopy, I wouldn't first have to install node and npm and less (and whatever browserify is), but could get away without it. :)

**Comment by wiredsister** [comment]
Re: @Engil 
1. I think browserify is a good solution. I'm definitely on board with only one JS file hitting the page.
2. Agreed.
3. Great. Let me know if I can help in any way.
4. Hm... allowing users to do customization via LESS files? You could do something like, have 3-4 themes that get applied when you have some sort of `.theme` class on the body. If you wanted an extension point for styles, we could make a blank LESS file to get loaded in lastly that we would point to them to for editing. We could also set up the `config.less` to be a bit more configurable. Another idea, is to provide different themes for Canopy that would be install-able somehow but not out of box. We'd have to provide documentation and pictures of different versions in an examples directory or something. It's really a UX question for Canopy. I'd be willing to do this work if there was some vision on what the themes would look like.
5. Sure. If that's what the kids want, tarball it up.

## 

While you didn't mention:
- compression/concatenation of CSS and JS
- post-processing of LESS for cross-browser support 
- linting for development
- hosting font locally (right now it's a google font)

I think these are also dependent on a consistent way we'll get packages and then call with some script to generate static assets.

**Comment by abbysmal** [comment]
Hello, thank you two for your replies. :)

> 3

Thank you! It's already on my local fork, I will push a branch in the next days and run a PR if you want to take a look at it.

> 4

Yeah the thing with theming is that the current situation on the HTML is a bit of a mess, you need to edit some HTML chunks at various parts of the code and it's not really the easiest thing to do.

I'm not sure about how far we can go with just providing CSS extension capabilities.
The thing is that on one side I'd like Canopy to be less attached to the blogging paradigm (that is, just some posts ordered by something like date or tags), on the other I wonder about how hard it would be. In my mind it would need some HTML templating to make something really customizable… I don't know how much we can rely on CSS to provide those kind of possibilities.

My first idea was something along the lines of making config.less a bit more configurable like you suggested, I think that's a good starting point :). The themes thing an interesting idea, but I'm still a bit lost on the HTML templating part and how to architecture it inside Canopy to make it easier to change how pages are structured, I think at some point that will be needed to allow truly interesting theming.

> compression/concatenation of CSS and JS

Yup, it would be really nice but I'm a bit lacking on the subject, do you have a set of tools in mind ? :)

> post-processing of LESS for cross-browser support

Also a bit lacking

> linting for development

That would be also nice, I'm not too sure where to plug it… in some sort of CI like travis ?
I need to explore what kind of funny things we can do with running stuff from `config.ml`. :)

> hosting font locally

Done by @voila (thank you again) the other, even though it looks like the weight of the font caused havoc inside @hannesm Canopy@Xen installation.

And about including JS/CSS file in the repository, I'll do that then, I need to finish migrating the style.sh to config.ml and that will come next. :)

**Comment by voila** [comment]
I will wait until JS/CSS are included before adding highlight.js (unless you have already done that somewhere...)

**Comment by wiredsister** [comment]
re @Engil: 
1. When I think CSS extention point theme, I think:

``` less
// =============
//   custom.less
// =============
  body {
    &.my-theme {
      background: @dark;
      color: @light;
      font-family: "Hacker sans";
    } 
    .some-widget {
      .my-theme & {
        float: left;
      }
    }
  }
```

If we're talking about changing layout significantly, then that's something else. Jekyll does that by using intermediate DSL like liquid lang to generate html snippets in order. If you want to order things by date or tag, that seems like a more fundamental change, so you're right more than CSS needs to be considered.
- compression: Tools in mind? No. But I know there are about a billion on NPM. I've used some before, forget names.
- linting, post-processing, firing the missles-- yeah we need to know more about config.ml capabilities. We totally could hook it into a Travis, but even just for local development to run before building, etc. 
- yay @voila

**Comment by hannesm** [comment]
@Engil I think this can be closed now (after #71) static assets are part of the git repo... everybody can tweak their css/js there...

**Comment by abbysmal** [comment]
Agreed
